src/smear/smear.c

```diff
@@ -169,6 +174,14 @@ EXPORT_SYMBOL void SRT_run(void)
 }
 
 EXPORT_SYMBOL void SRT_wait_for_idle(void)
+{
+    sem_wait(&idle_sem);
+    flushEventQueue();
+    sem_post(&idle_sem);
+    return;
```

Comment:
Isn't it enough to just call `sem_wait` then `sem_post`?  Two reasons:
* Although the queue may currently have something in it, it was recently deemed idle by `mainloop`, which I would think should be good enough, and
* `SRT_wait_for_idle` is almost definitely called from a different thread than `mainloop`, so although the queue is protected, that's asking for lock contention in a common case.  It also means that there could be behavior in the handler that get performed across different threads (inside a lock, granted, but this could still be problematic).

Also, if for some stupid reason `SRT_wait_for_idle` is called within a handler, this will deadlock.

Reply:
That's what I tried first, but there were bugs - it didn't usually cause anything to happen. Now that I think about it, though, the bug was probably related to the change I made in `flushEventQueue`. 

Let me try it, and see if it's better now. I think it still deadlocks if it's called inside an event handler, but...I'll put a comment in to say not to do that.

Reply:
Nope, tried it and it's no good. Here's why:

The main thread schedules a bunch of events, then calls `SRT_wait_for_idle`. We expect this to run until there are no immediately pending events, but instead it just pops in, sees that it can have the semaphore, and pops back out. 0 events processed.

Incidentally, both functions can deadlock if called from inside event handlers. I added a warning in the header about it.

Reply:
Aha!  I think we can avoid that by setting the initial value to 0 and doing the following

    static void *mainloop(void *unused)
    {
        while(true)
        {
            flushEventQueue();
            sem_post(&idle_sem);
            sem_wait(&idle_sem);
            // We might not even need the sleep(0) here anymore.
        }
        return NULL;
    }

    EXPORT_SYMBOL void SRT_wait_for_idle(void)
    {
        sem_wait(&idle_sem);
        sem_post(&idle_sem);
    }

With this order, if the main thread enters `SRT_wait_for_idle`, it will block until `flushEventQueue()` is finished.  If instead the thread yields to `mainloop` without taking the semaphore until after it calls `sem_post`, then that's safe; if it misses that too, then it degenerates to the first case.  Also, as I said in the comment, I think that with the semaphore there we no longer need a sleep, since semaphores are guaranteed to transition to anything waiting.  Either way `mainloop` is chewing up its entire CPU busy-waiting on an empty queue, but we already had that problem.